### ISSUE TITLE
SparseCategoricalCrossentropy and Mixed Precision Training

### ISSUE THREAD

**Issue by zuyezheng** [keras-team/keras#15012]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary):  Binary
- TensorFlow version (use command below): v2.5.0-0
- Python version: 3.8
- Bazel version (if compiling from source):
- GPU model and memory: A6000
- Exact command to reproduce:

tf.keras.losses.SparseCategoricalCrossentropy()

**Describe the problem**.

`sparse_categorical_crossentropy` in `losses.py` performs an unnecessary [cast](https://github.com/tensorflow/tensorflow/blob/8c541bab667488249a34404170985dff8a9bab56/tensorflow/python/keras/losses.py#L1751) of `y_true` to `y_pred.dtype` since it's then [cast](https://github.com/keras-team/keras/blob/d3688b72924a4235598f0f80038de8c897f44799/keras/backend.py#L4952) to `int64` in `sparse_categorical_crossentropy` in `keras.backend.py`. Eventual call to `sparse_softmax_cross_entropy_with_logits` in `nn_ops.py` is documented to expect int64 as well.

This seems to be the same code as in `categorical_crossentropy`, but causes issues with sparse, especially with mixed precision training and float16 as the loss in precision causes incorrect encodings or labels outside the domain resulting in incorrect or `nan` loss. With float16, issues start with a couple thousand labels and a couple hundred labels with bfloat16.

**Describe the current behavior**.

Loss of precision for labels.

**Describe the expected behavior**.

Cast of `y_true` to `y_pred.dtype` should be skipped.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

https://colab.research.google.com/drive/1oRbNOnCo1i2HcXD2V4_-D1Bz2EVxiT65

**Comment by rmothukuru** [comment]
@zuyezheng,
Similar issue has been raised in #15014 with **`other losses`** and a [**`PR`**](https://github.com/keras-team/keras/pull/15015#issue-698536884) also has been raised. Can we close this issue so that we can track it in #15014? Thanks!

**Comment by zuyezheng** [comment]
@rmothukuru ah thanks, did a search in tf repo and didn't find it, forgot to do one here.
